be/src/storage/rowset/dict_page.h

```diff
@@ -0,0 +1,162 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "gen_cpp/segment.pb.h"
+#include "gutil/hash/string_hash.h"
+#include "runtime/mem_pool.h"
+#include "storage/olap_common.h"
+#include "storage/range.h"
+#include "storage/rowset/bitshuffle_page.h"
+#include "storage/rowset/common.h"
+#include "storage/rowset/options.h"
+#include "storage/rowset/plain_page.h"
+#include "storage/types.h"
+#include "util/phmap/phmap.h"
+
+namespace starrocks {
+
+// This type of page use dictionary encoding for numbers.
+// There is only one dictionary page for all the data pages within a column.
+//
+// Layout for dictionary encoded page:
+// Either header + embedded codeword page, which can be encoded with any
+//        int PageBuilder, when mode_ = DICT_ENCODING.
+// Or     header + embedded BitshufflePageBuilder, when mode_ = PLAIN_ENCOING.
+// Data pages start with mode_ = DICT_ENCODING, when the the size of dictionary
+// page go beyond the option_->dict_page_size, the subsequent data pages will switch
+// to string plain page automatically.
+
+// DictPageBuilder has two encoders, data-page-builder and dict-builder
+// dict-builder is used to encode the dictionary, data-page-builder is used to encode the
+// dictionary's index. data-page-builder and dict-builder use BitshufflePageBuilder for encoding
+// Because when the dictionary page is full, data-page-builder will no longer store the index of the
+// dictionary page, but instead store the data itself, data-page-builder needs to reserve a segment
+// of space in advance to store the encoding type, indicating whether this page stores the index of
+// the dictionary page or the data
+
+// The maximum value of int32 is 2147483647, representing that a dictionary page can store
+// a maximum of 2147483648 elements. If a dictionary page stores 2147483648 elements, assuming
+// one element only occupies one byte, the size of the dictionary page will be 2GB, which is
+// impossible. Therefore, int32 is sufficient to store the index of the dictionary page,
+// and overflow will not occur.
+template <LogicalType field_type>
+struct DataTypeTraits {
+    static const LogicalType type = TYPE_INT;
+};
+
+template <>
+struct DataTypeTraits<TYPE_SMALLINT> {
+    static const LogicalType type = TYPE_UNSIGNED_SMALLINT;
```

Comment:
UNSIGNED?

Reply:
Yes, it should be UNSIGNED. The meaning is TYPE_SMALLINT value type , the dictionary key type is TYPE_UNSIGNED_SMALLINT.